service/history/queue/cross_cluster_queue_processor_base.go

```diff
@@ -0,0 +1,688 @@
+// Copyright (c) 2017-2021 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package queue
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
+	"github.com/uber/cadence/common/types"
+	"github.com/uber/cadence/service/history/config"
+	"github.com/uber/cadence/service/history/shard"
+	"github.com/uber/cadence/service/history/task"
+)
+
+type (
+	crossClusterTaskKey = transferTaskKey
+
+	crossClusterQueueProcessorBase struct {
+		sync.Mutex
+
+		*processorBase
+		targetCluster   string
+		taskInitializer task.Initializer
+
+		processingLock sync.Mutex
+		backoffTimer   map[int]*time.Timer
+		shouldProcess  map[int]bool
+
+		outstandingTaskCount int
+
+		notifyCh  chan struct{}
+		processCh chan struct{}
+	}
+)
+
+func newCrossClusterQueueProcessorBase(
+	shard shard.Context,
+	targetCluster string,
+	processingQueueStates []ProcessingQueueState,
+	taskProcessor task.Processor,
+	options *queueProcessorOptions,
+	updateMaxReadLevel updateMaxReadLevelFn,
+	updateProcessingQueueStates updateProcessingQueueStatesFn,
+	queueShutdown queueShutdownFn,
+	taskExecutor task.Executor,
+	logger log.Logger,
+	metricsClient metrics.Client,
+) *crossClusterQueueProcessorBase {
+	processorBase := newProcessorBase(
+		shard,
+		processingQueueStates,
+		taskProcessor,
+		options,
+		updateMaxReadLevel,
+		nil,
```

Comment:
It will panic if `EnablePersistQueueStates ` is false.

Reply:
Yes. Set it with a static value true